ZOOKEEPER-2939: Added last/min/max proposal size JMX beans

anmolnar opened the pull request:
https://issues.apache.org/jira/browse/ZOOKEEPER-2939
- Refactor proposal serialization logic to a common place (SerializeUtils),
- Add JMX metric to monitor jute.maxbuffer setting
- Add JMX metric to monitor min/max/last size of generated proposals
- Add proposal stats info to 'stat' 4lw command
- Add proposal stats resetting to 'srst' 4lw command
- Unit tests

afine commented on src/java/main/org/apache/zookeeper/server/quorum/LeaderStats.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Size of the largest proposal which has been generated since the server was started.
+     */
+    private int maxProposalSize = 0;
```
nit: can we set all of these to -1 for consistency

anmolnar replied:
Done.

anmolnar commented on src/java/main/org/apache/zookeeper/server/ZooKeeperServerBean.java:
```diff
@@ -167,4 +167,9 @@ public String getSecureClientAddress() {
+
+    @Override
+    public int getJuteMaxBufferSize() {
+        return Integer.getInteger("jute.maxbuffer", 0xfffff);
```
I deal with it in the other way around. Directly access BinaryInputArchive.maxBuffer from here.

anmolnar commented on src/java/main/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -34,12 +33,12 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
```
Good catch, thanks.

anmolnar commented on src/java/main/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -984,8 +990,6 @@ public void commitAndActivate(long zxid, long designatedLeader) {
     /**
      * Create an inform packet and send it to all observers.
-     * @param zxid
-     * @param proposal
```
Intellij keep notifying me about these comments not referring to valid parameters in the method definition anymore. I thought it'd be nice to do some cleanup together with the patch.

anmolnar commented on src/java/main/org/apache/zookeeper/server/quorum/LeaderStats.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * Provides live statistics about a running Leader.
+ */
+class LeaderStats {
```
Done.

anmolnar commented on src/java/test/org/apache/zookeeper/server/quorum/LeaderStatsTest.java:
```diff
@@ -0,0 +1,39 @@
+
+public class LeaderStatsTest {
+    @Test
+    public void testSetProposalSize_SetMinMax() {
```
Removed.

anmolnar commented on src/java/test/org/apache/zookeeper/server/quorum/LeaderStatsTest.java:
```diff
@@ -0,0 +1,39 @@
+public class LeaderStatsTest {
+    @Test
+    public void testSetProposalSize_SetMinMax() {
+        LeaderStats stats = new LeaderStats();
```
Good point. I've also added checking the lastProposalSize size value besides min/max.

anmolnar commented on src/java/test/org/apache/zookeeper/server/util/SerializeUtilsTest.java:
```diff
@@ -0,0 +1,73 @@
+        Request request = new Request(1, 2, 3, header, txn, 4);
+        byte[] data = SerializeUtils.serializeRequest(request);
+        assertNotNull(data);
+        InOrder inOrder = inOrder(header, txn);
```
Very good question, I'd been thinking about it for a while. The reason why I eventually left it out is that this unit test verifies the behaviour of serializeRequest() method and _not_ the serialization itself. So basically we have to verify that the right serialize() methods get called in the right order instead of what they actually do. 

On the other hand, you're right, because if we mock the serialize() methods, they won't have sideeffect on our test.

What do you think?

anmolnar commented:
@afine I've created 2 subtasks for the Jira as you suggested and changed this PR to refer to the right one.

anmolnar commented:
@afine Take a look at please the latest commit, I've added verification of the byte array itself to serializeRequest() unit test. I'm not 100% sure it's needed, but it might makes sense this way.

anmolnar commented:
@afine Thanks for approving.
I've added the same byte array verification to the other unit test too.
Also squashed everything into single commit.

Ready to merge.

phunt commented:
I didn't look at this in much depth yet but one obvious things struck me - you are missing the reset capability. See how we handle latency tracking in the beans.

We try to keep jmx, 4lw and jetty interfaces (new in 3.5) in lock step wrt the metrics they provide. What do you think, do you want to try to include that in this patch or create separate JIRA(s) to track/address?

anmolnar commented:
@phunt Thanks for looking into this. I'll include the reset capability into this patch as suggested.

anmolnar commented:
@phunt I've added capability of resetting proposal statistics as well as unit tests for LeaderBean class.

phunt commented:
What's the plan re my other comment

> We try to keep jmx, 4lw and jetty interfaces (new in 3.5) in lock step wrt the metrics they provide.

anmolnar commented:
@phunt Sorry, I've completely forgotten about it.
I believe the easiest would be to add proposal stats to 'stat' 4lw in this patch, so that it could be easily applied to all major branches.
After that I'd create a separate PR for jetty changes which will target trunk and 3.5 only.

anmolnar commented:
@phunt 4lw changes have been added.
@afine please start the review over, because so many things have been changed since you approved it.

For some weird reason I cannot get a green build, but seems like it's unrelated to this patch.

afine commented on src/java/test/org/apache/zookeeper/server/quorum/StatCommandTest.java:
```diff
@@ -0,0 +1,106 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.hamcrest.core.StringContains.containsString;
```
According to: https://github.com/junit-team/junit4/wiki/matchers-and-assertthat we should probably use `import static org.hamcrest.CoreMatchers.containsString;`

In addition, this will not resolve in intellij until we allow transitive test dependencies to be imported by changing build.xml:1820

anmolnar replied:
Thanks for the investigation. All done.

anmolnar commented:
@afine You're not missing anything, I didn't add jetty support to this patch intentionally, because afaik jetty is not supported in 3.4. So the plan is to keep this patch compatible with all branches and I'll create a separate one for 3.5/trunk only.

anmolnar commented on src/java/test/org/apache/zookeeper/server/quorum/StatResetCommandTest.java:
```diff
@@ -0,0 +1,112 @@
+import static org.mockito.Mockito.when;
+
+public class StatResetCommandTest {
+    private static final String ZK_NOT_SERVING = "This ZooKeeper instance is not currently serving requests\n";
```
Done.

afine commented on build.xml:
```diff
@@ -1817,7 +1817,7 @@ xmlns:cs="antlib:com.puppycrawl.tools.checkstyle.ant">
        <ivy:resolve useOrigin="true" conf="test"/>
        <ivy:cachepath pathid="default.path.id" conf="default" />
-       <ivy:cachepath pathid="junit.path.id" conf="test" transitive="false"/>
+       <ivy:cachepath pathid="junit.path.id" conf="test" transitive="true"/>
```
nit: we can just remove the transitive property to be consistent

anmolnar commented:
@afine @phunt Findbugs issues have been resolved. Please review & commit.

anmolnar commented:
@phunt do you think we can commit this patch?

anmolnar commented:
@phunt @afine are u happy with the change?

phunt commented:
This seems like a reasonable addition (see below). I like the refactorings (kudos). The code seems fine. I do have a couple questions though.

1) this information should also be added to mntr command output. The idea of mntr is to have machine readable output - java properties format (we had this prior to adding the jetty support).

2) the description for ZOOKEEPER-2939 is missing. I think it would be good to give some insight into why you are proposing/submitting this change. Esp around operational aspect.

3) it's not clear to me how useful this will be in practice, from an operations perspective. Hopefully the answer to item 2 here will help. But say you are running a cluster in production. It will be under some load, with new proposals showing up pseudo-randomly, perhaps rapidly. last and min seem of very little use (which perhaps is fine, similar to us capturing min latency). Max I can see as useful. Why did you not decide to include "avg" for example? Unfortunately our stats tracking was created before some of the metrics tracking libraries available today were generally available. (might be a good project for someone to investigate separately from this PR).

anmolnar commented:
@phunt That's a perfectly valid point Pat, I've been asked about it several times before. Therefore I added some more comments to the parent Jira of monitoring, please take a look at the description: https://issues.apache.org/jira/browse/ZOOKEEPER-2933

I've also populated the description of subtasks too to help understanding them.

Regarding to statistics tracking in Zookeeper: I wonder how much does it make sense to implementing something complicated within the server. One approach (which this PR tries to follow) is to expose basic values which can be sampled by a more sophisticated monitoring tool. Average, sliding window, etc. can be implemented in there. 

On the other hand we could add some more clever logic to ZK as well such as 
https://google.github.io/guava/releases/21.0/api/docs/com/google/common/math/Quantiles.html
or
https://commons.apache.org/proper/commons-math/javadocs/api-3.0/org/apache/commons/math3/stat/descriptive/rank/Percentile.html
Basically calculating percentiles (50-90-99-Median-Max-etc.) with the appropriate library and expose the values with JMX. This will probably require us to gather samples with a sliding window within Zookeeper.

anmolnar commented:
@phunt I added proposal stats to 'mntr' command output as requested.

In the meantime I found the following metrics lib which could be useful and easily integrated into Zookeeper:

http://metrics.dropwizard.io/3.2.3/getting-started.html

Do you think it's worth merging this PR and refactor it in a separate one to use that lib?

phunt commented:
> Do you think it's worth merging this PR and refactor it in a separate one to use that lib?

I'd say, let's consider it separately. There's probably some work to do to identify what we should go to. IIRC at a recent meetup the Facebook folks mentioned they have done extensive work to add more metrics. We might look at what they have done. Similarly other Apache projects in the Hadoop ecosystem have also moved on, we might look at what they are using (although if I had to guess it's probably dropwizard, but I'm not sure)

phunt commented:
> One approach (which this PR tries to follow) is to expose basic values which can be sampled by a more sophisticated monitoring tool. Average, sliding window, etc. can be implemented in there.

That's fine, but I guess my thought was that sampling in this context will have very low fidelity vs having the server calculate, the avg for example. Given the monitoring tool will poll every 5 seconds or so (perhaps less frequently) the number of samples vs the number of proposals could be very low. Not a reason not to do it, but I'm not sure you're really going to get what you expect as an operator.

anmolnar commented:
@phunt I asked around quickly:
HBase - own library (looks like they've started to use dropwizard somehow)
HDFS - Hadoop Metrics2 (included in hadoop-common)
Kafka - Yammer (dropwizard predecessor)
Cassandra - Yammer

I'll give dropwizard a try first.

anmolnar commented:
@phunt @afine Do you think this PR is ready to committed before the Dropwizard one?

afine commented:
@phunt I think this is ready to merge. Just waiting on your +1.
